### PR TITLE
Adding Dispose to browser

### LINKED ISSUE #50: Browser.cs needs to implement IDisposable
Long story short, when using Selenate with TestFramework, Selenium doesn't get cleaned up correctly because, to date, we've cleaned up Selenium ourselves. Using these two together should be easier to cleanup... I.E. implement IDisposable and do a driver.Dispose.

### PR DESCRIPTION
**Pull request by PandaMagnus**
Resolves #50

### CODE REVIEW

**Comment by PandaMagnus** [comment]
I didn't consider what would happen if the class is derived from (which, while I personally haven't hit a use case, I can definitely imagine some scenarios where that could occur). I'll go ahead and update to follow the BaseClass example.

**Comment by PandaMagnus** [comment]
> @Keboo I didn't consider what would happen if the class is derived from (which, while I personally haven't hit a use case, I can definitely imagine some scenarios where that could occur). I'll go ahead and update to follow the BaseClass example.

@Keboo It's not clear to me from the documentation, would implementing my own SafeHandle be a good option here? That seems weird to me, since really I just need to clean up Selenium in a way that common patterns understand (e.g. if they invoke Browser.cs in a using statement, or if they're using a service provider, or etc.)

**Comment by Keboo** [comment]
Was picturing something much simpler like this:

```
protected virtual void Dispose(bool disposing)
{
    if (disposing)
    {
        Driver?.Dispose();
    }
}

public void Dispose()
{
    Dispose(true);
    GC.SuppressFinalize(this);
}
```

**Comment by Keboo** [comment]
Was picturing something much simpler like this:

```
protected virtual void Dispose(bool disposing)
{
    if (disposing)
    {
        Driver?.Dispose();
    }
}

public void Dispose()
{
    Dispose(true);
    GC.SuppressFinalize(this);
}
```

**Comment by PandaMagnus** [comment]
@Keboo Oh dude, that's way simpler. Okay cool, I can deal with that. :) Do you still think it's worth saving off whether it's been disposed or not and just returning if it has?

**Comment by Keboo** [comment]
Yea, you can if you want. The key details is that calling Dispose multiple times should be handled. Exactly how that gets done is an implementation detail.

**Comment by PandaMagnus** [comment]
@Keboo Done and done (I think. :) )

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs [hunk 1]
```diff
@@ -189,6 +189,20 @@ public void FrameSwitchAttempt(int secondsToWait = 15, params By[] bys)
...
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Has Selenium already been disposed?
+        /// </summary>
+        protected bool Disposed = false;
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]: If you are going to track your disposed state this should be a private field.
```suggestion
        private bool _Disposed;
```

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs [hunk 2]
```diff
@@ -252,5 +266,24 @@ protected IWebDriver InitDriver(BrowserType browser)
...
             return driver;
         }
+
+        /// <summary>
+        /// Disposes the current Selenium Driver
+        /// </summary>
+        /// <param name="disposing">Did the call come from Dispose()?</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if(Disposed)
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]:
```suggestion
            if(_Disposed)
```

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs [hunk 3]
```diff
@@ -252,5 +266,24 @@ protected IWebDriver InitDriver(BrowserType browser)
...
+            {
+                return;
+            }
+
+            if(disposing)
+            {
+                Driver?.Dispose();
+            }
+
+            Disposed = true;
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]:
```suggestion
            _Disposed = true;
```

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs [hunk 4]
```diff
@@ -189,6 +189,20 @@ public void FrameSwitchAttempt(int secondsToWait = 15, params By[] bys)
...
 
+        /// <summary>
+        /// Disposes the current Selenium Driver
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]:
```suggestion
   ~Browser()
   {
      Dispose(false);
   }
```

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs (hunk 1 above)
Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]: Understood, thank you!

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs (hunk 2 above)
Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs]: Oops, missed this when committing your suggestions. Ended up catching it and changing it in a follow-up commit.

**Pull request merged**